MudDataGrid: Allow user to use a custom IComparer<object> for sorting

SinisterMaya opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->


## Description
<!-- Describe your changes in detail any why -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310 -->

Proposition to add a IComparer to the SortDefinition record class in DataGrid. User can either set their custom Comparer as a Column Parameter or they can set it once with SetSortAsync Task or the SortDefinitions Dictionary.

This feature is optional to be used. It's a tool to allow advanced sorting. Those wishing to use a custom comparer have to implement their own Comparer class like the one I made for my example and my test unit.
![image](https://user-images.githubusercontent.com/29356866/220725358-d58b946d-b637-42ef-886e-d13373d00cbe.png)

The reason why I did this was to allow custom Comparer like the StrCmpLogicalW from the Shlwapi.dll on Windows so we can properly sort filenames or string with a hierarchical pattern (e.g.: {"1", "1.1", "1.1.2", "1.2", "1.10"})

Since OrderBy and OrderByDescending both takes a nullable IComparer<TKey> parameter, we can default it as null without causing a breaking change. That way, if comparer is not supplied, OrderBy and OrderByDescending will use the default comparer for the type.

Here's the proposed documentation addition with it:
(default string sorting)
![image](https://user-images.githubusercontent.com/29356866/220721736-14265fb1-f0e2-4a47-8cd0-ac2eafd29184.png)

(natural sorting enabled by the Comparer Parameter)
![image](https://user-images.githubusercontent.com/29356866/220721864-1d54752e-c1fc-43f7-bd5d-abee4cb6b88e.png)

We can see that 10 comes now after 2 and 1_2 comes now before 1_10 and 1_11

Ordering using a IComparer is not supported for Linq to Entities operations, but I don't think that is ever the case here unless a Column can sort directly from a Linq query ?

## How Has This Been Tested?
<!-- All PR's should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->

I created a unit test to validate calling SetSortAsync works with and without the comparer on different field types. I also tested assigning a comparer to a column and validated that the ordering was the one expected.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

Commit:
MudDataGrid - SortDefinition: Added a IComparer<object> to allow user to add custom Comparer

Anu6is commented on src/MudBlazor.Docs/Pages/Components/DataGrid/DataGridPage.razor:
```diff
@@ -116,6 +116,20 @@
+        <DocsPageSection>
+            <SectionHeader Title="Advanced Sorting">
+                <Description>
+                    The <CodeInline>&lt;MudDataGrid></CodeInline> allows you to provide your own <CodeInline>Comparer</CodeInline> for To allow advanced sorting. Keep in mind that the <CodeInline>Comparer</CodeInline>
```
There's a typo with this line "your own <CodeInline>Comparer</CodeInline> **for To allow** advanced sorting."

should be either 
your own Comparer **for** advanced sorting
or 
your own Comparer **to allow** advanced sorting

SinisterMaya replied:
Oh, many thanks! Will correct it

Anu6is reviewed (changes requested):
Small correction in documentation

SinisterMaya commented:
EDIT:

Corrected my typo (Thank you @Anu6is) and added ctrl/alt click tests to my unit test

henon commented:
DataGrid had a huge update so please merge dev and resolve conflicts.

SinisterMaya commented:
@henon I rebased and merged. The conflict was the change from Column.Field to Column.PropertyName when handling sort change on header cells on lines where I added a comparer to the SortDefinition. This doesn't break anything for me. It was just a matter of combining changes.

I also had to change my test and doc datagrid to adopt the new PropertyColumn replacing Column.

henon reviewed (changes requested):
I think the natural comparison is great but it makes the example quite long. Also it seems duplicated in example and test case. So how about moving it into a public MudBlazor util class which is used by both. Public so it can also be used by our users.

SinisterMaya commented:
@henon 

I can do that. It was duplicated because I thought of it as an example. We can roughly reach the same result with the Shell windows api dll on windows: 
![image](https://user-images.githubusercontent.com/29356866/225120110-d4c2fa45-0a83-46a7-8a15-ebbf48cd54d5.png)

But since it had to work without any references, I used this algorithm I found which is about twice as slow, but still faster than a regex approach. I've been working on optimizing it, but did not yield huge results so far. Here's the Benchmark of it.
![image](https://user-images.githubusercontent.com/29356866/225122149-d006b1dc-43f9-45f6-a361-98b8729d24fa.png)

With this being the regex algorithm in my benchmark:
![image](https://user-images.githubusercontent.com/29356866/225123237-4cec67fa-a33a-4255-b93b-b81c27d8bb5a.png)

Commit:
MudDataGrid - SortDefinition: Added NaturalComparer in Utilities and created unit tests for.

SinisterMaya commented:
Changes added in a new commit. Figured it would help keep track if requested changes are good for you. I did a UnitTest to test pretty much every lines of the NaturalComparer Utility. Changed my examples to use the public utility instead of their own comparer.

henon commented on src/MudBlazor.UnitTests/Utilities/NaturalComparerTest.cs:
```diff
@@ -0,0 +1,406 @@
+            ,"1.25.txt"
+            ,"1.txt"
+            ,"1a.txt"
+            ,"111111111111111111.txt"
+            ,"00222222222222222222.txt"
+            ,"0222222222222222222.txt"
+            ,"222222222222222222.txt"
+            ,"00999999999999999999.txt"
+            ,"0999999999999999999.txt"
+            ,"999999999999999999.txt"
+            ,"1111111111111111111.txt"
+            ,"2222222222222222222.txt"
+            ,"9999999999999999999.txt"
+            ,"11111111111111111111.txt"
+            ,"0001111111111111111111111111111111111111111111111111111111111111111111.txt"
+            ,"1111111111111111111111111111111111111111111111111111111111111111111.txt"
+            ,"9999999999999999999999999999999999999999999999999999999999999999999.txt"
+            ,"11111111111111111111111111111111111111111111111111111111111111111111.txt"
```
not a critique, just a fun fact. I tried the windows api dll func on my win10 to sort these file names and it sorts them differently. Very subtle differences though.

```
"1-a.txt",
"0001111111111111111111111111111111111111111111111111111111111111111111.txt",
"00222222222222222222.txt",
"00999999999999999999.txt",
"111111111111111111.txt",
"0222222222222222222.txt",
"222222222222222222.txt",
"0999999999999999999.txt",
"999999999999999999.txt",
"1111111111111111111.txt",
"11111111111111111111.txt",
"1111111111111111111111111111111111111111111111111111111111111111111.txt",
"11111111111111111111111111111111111111111111111111111111111111111111.txt",
"2222222222222222222.txt",
"9999999999999999999.txt",
"9999999999999999999999999999999999999999999999999999999999999999999.txt",
"a001test01.txt",
```

Hard to say which is better.

SinisterMaya replied:
Yeah. Apparently, Windows API sorting differs from versions to versions and it's hard to tell how they evaluate non alphanumerical char weight. We can see that windows API doesn't give"-" the same weight as "(", "[", "_", ... etc. "-" seems to be ignored by windows, which seems to be inconsistent with what I would expect. what do you think?

![image](https://user-images.githubusercontent.com/29356866/225299430-1e9f8efe-0c36-4c03-aad6-78c3dc0c1e7c.png)

henon replied:
yeah, these are all special cases, so don't worry about it. for most people what really matters is the order of 1, 2, 3, 10, 11, 12 instead of 1, 10, 11, 12, 2, 3

SinisterMaya replied:
yes. in my case (and the reason why I created that PR) is to order hierarchical item name 

![image](https://user-images.githubusercontent.com/29356866/225311334-4ab54f93-0f03-4adb-bf29-ff43a15b9398.png)

This project represents a Bill Of Material. Each Item is either a part to be manufactured, ordered, or an assembly/sub-assembly. You can Imagine a keyboard as being a main assembly, each keyswitches are subassemblies containing their own components and so on. Some component are found in different subassemblies at different levels. So my items are a way to hold that information while the actual component (The Model) can be referenced more than once.

henon commented:
@tjscience I think adding support for a comparer is a good idea. shall I merge this now or do you want me to merge it later to avoid possible conflicts with your changes?

neozhu commented:
Can the muddatagrid component support remote querying, filtering, and sorting on the backend api?

henon commented:
@neozhu you should post this in its own issue and tag @tjscience. 

@SinisterMaya could you please resolve the conflicts? we recently merged a lot of datagrid PRs. I think I'll merge yours now.

SinisterMaya commented:
@henon I rebased and merged. It was only the scoped nullable context that was added to the SortDefinition.cs file

henon commented:
Thanks a lot! 

@tjscience FYI
